cloud/google/serviceaccount.go

```diff
@@ -26,68 +23,142 @@ import (
+
+var (
+	MasterNodeRoles = []string{
+		"compute.instanceAdmin",
+		"compute.networkAdmin",
+		"compute.securityAdmin",
+		"compute.viewer",
+		"iam.serviceAccountUser",
+		"storage.admin",
+		"storage.objectViewer",
+	}
+	WorkerNodeRoles = []string{
+		"compute.instanceAdmin",
+		"compute.networkAdmin",
+		"compute.securityAdmin",
+	}
+	MachineControllerRoles = []string{
+		"compute.instanceAdmin.v1",
+		"iam.serviceAccountActor",
+	}
 )
 
+// Returns the email address of the service account that should be used
+// as the default service account for this machine
+func (gce *GCEClient) GetDefaultServiceAccountForMachine(cluster *clusterv1.Cluster, machine *clusterv1.Machine) string {
+	if util.IsMaster(machine) {
+		return cluster.ObjectMeta.Annotations[ClusterAnnotationPrefix+MasterNodeServiceAccountPrefix]
+	} else {
+		return cluster.ObjectMeta.Annotations[ClusterAnnotationPrefix+WorkerNodeServiceAccountPrefix]
+	}
+}
+
+// Creates a GCP service account for the master node, granted permissions
+// that allow the control plane to provision disks and networking resources
+func (gce *GCEClient) CreateMasterNodeServiceAccount(cluster *clusterv1.Cluster, initialMachines []*clusterv1.Machine) error {
+	_, _, err := gce.createServiceAccount(MasterNodeServiceAccountPrefix, MasterNodeRoles, cluster, initialMachines)
+
+	return err
+}
+
+// Creates a GCP service account for the worker node
+func (gce *GCEClient) CreateWorkerNodeServiceAccount(cluster *clusterv1.Cluster, initialMachines []*clusterv1.Machine) error {
+	_, _, err := gce.createServiceAccount(WorkerNodeServiceAccountPrefix, WorkerNodeRoles, cluster, initialMachines)
+
+	return err
+}
+
 // Creates a GCP service account for the machine controller, granted the
 // permissions to manage compute instances, and stores its credentials as a
 // Kubernetes secret.
 func (gce *GCEClient) CreateMachineControllerServiceAccount(cluster *clusterv1.Cluster, initialMachines []*clusterv1.Machine) error {
+	accountId, project, err := gce.createServiceAccount(MachineControllerServiceAccountPrefix, MachineControllerRoles, cluster, initialMachines)
+	if err != nil {
+		return err
+	}
+
+	email := fmt.Sprintf("%s@%s.iam.gserviceaccount.com", accountId, project)
+
+	localFile := accountId + "-key.json"
+	err = run("gcloud", "--project", project, "iam", "service-accounts", "keys", "create", localFile, "--iam-account", email)
+	if err != nil {
+		return fmt.Errorf("couldn't create service account key: %v", err)
+	}
+
+	err = run("kubectl", "create", "secret", "generic", MachineControllerSecret, "--from-file=service-account.json="+localFile)
+	if err != nil {
+		return fmt.Errorf("couldn't import service account key as credential: %v", err)
+	}
 
+	if err := run("rm", localFile); err != nil {
+		glog.Error(err)
+	}
+
+	return nil
+}
+
+func (gce *GCEClient) createServiceAccount(serviceAccountPrefix string, roles []string, cluster *clusterv1.Cluster, initialMachines []*clusterv1.Machine) (string, string, error) {
```

Comment:
What's the `string, string, error` return values? It's not obvious without digging through the implementation.